csi/controller_server.go

```diff
@@ -545,6 +547,41 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 		return nil, status.Error(codes.InvalidArgument, "Snapshot name must be provided")
 	}
 
+	volume, err := cs.apiClient.Volume.ById(csiVolumeName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get volume %v", csiVolumeName)
+	}
+	nodeID := ""
+	if volume.State == string(longhorn.VolumeStateDetached) {
+		nodeID, err = cs.autoAttachVolumeToRandomNode(volume)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+	if nodeID != "" {
+		defer func() {
```

Comment:
```suggestion
	if volume.State == string(longhorn.VolumeStateDetached) {
		nodeID, err := cs.autoAttachVolumeToRandomNode(volume)
		if err != nil {
			return nil, status.Error(codes.Internal, err.Error())
		}
		defer func() {
			// auto detachment
			......
		}
	}
```

I am not sure why you prefer to use `nodeID` as the indication of volume auto attachment/detachment.

Reply:
Fixed.